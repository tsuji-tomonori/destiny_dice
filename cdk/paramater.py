from pathlib import Path

import tomllib

with (Path.cwd() / "pyproject.toml").open("rb") as f:
    project = tomllib.load(f)["project"]["name"]


paramater = {
    "lambda": {
        "create_pool": {
            "env": {
                "LOG_LEVEL": "INFO",
            },
            "memory_size": 128,
        },
        "list_pool": {
            "env": {
                "LOG_LEVEL": "INFO",
            },
            "memory_size": 128,
        },
        "delete_pool": {
            "env": {
                "LOG_LEVEL": "INFO",
            },
            "memory_size": 128,
        },
        "dice": {
            "env": {
                "LOG_LEVEL": "INFO",
            },
            "memory_size": 128,
        },
    },
    "waf": {
        "common": {
            "addresses": [],
            "ip_address_version": "IPV4",
            "metric_name": project,
            "priority": 100,
        },
        "apigw": {
            "scope": "REGIONAL",
        },
    },
}


def build_name(service: str, hostname: str) -> str:
    return f"dice-{service}-{hostname}"
