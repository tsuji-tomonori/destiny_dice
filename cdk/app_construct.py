from typing import Any, Self

from aws_cdk import aws_apigateway as apigw
from constructs import Construct

from cdk.infra_construct import InfraConstruct
from cdk.lmd_construct import LambdaConstruct
from cdk.paramater import build_name


class AppConstruct(Construct):
    def __init__(
        self: Self,
        scope: Construct,
        construct_id: str,
        infra: InfraConstruct,
        **kwargs: Any,  # noqa: ANN401
    ) -> None:
        super().__init__(scope, construct_id, **kwargs)

        self.api = apigw.RestApi(
            scope=self,
            id="api",
            default_cors_preflight_options=apigw.CorsOptions(
                allow_origins=apigw.Cors.ALL_ORIGINS,
                allow_methods=apigw.Cors.ALL_METHODS,
            ),
            rest_api_name=build_name("api", "destiny_dice"),
            description="destiny_dice",
        )
        pools = self.api.root.add_resource("pools")
        pool_name = pools.add_resource("{pool_name}")
        dice = pool_name.add_resource("dice")

        self.create_pool = LambdaConstruct(self, "create_pool", infra)
        pool_name.add_method(
            http_method="POST",
            integration=apigw.LambdaIntegration(
                handler=self.create_pool.function,
            ),
        )
        assert self.create_pool.function.role is not None
        infra.table_pool.grant_read_write_data(self.create_pool.function.role)
        self.create_pool.function.add_environment(
            key="POOL_TABLE_NAME",
            value=infra.table_pool.table_name,
        )
        infra.table_item.grant_read_write_data(self.create_pool.function.role)
        self.create_pool.function.add_environment(
            key="ITEM_TABLE_NAME",
            value=infra.table_item.table_name,
        )

        self.list_pool = LambdaConstruct(self, "list_pool", infra)
        pools.add_method(
            http_method="GET",
            integration=apigw.LambdaIntegration(
                handler=self.list_pool.function,
            ),
        )
        assert self.list_pool.function.role is not None
        infra.table_pool.grant_read_data(self.list_pool.function.role)
        self.list_pool.function.add_environment(
            key="POOL_TABLE_NAME",
            value=infra.table_pool.table_name,
        )

        self.delete_pool = LambdaConstruct(self, "delete_pool", infra)
        pool_name.add_method(
            http_method="DELETE",
            integration=apigw.LambdaIntegration(
                handler=self.delete_pool.function,
            ),
        )
        assert self.delete_pool.function.role is not None
        infra.table_pool.grant_read_write_data(self.delete_pool.function.role)
        self.delete_pool.function.add_environment(
            key="POOL_TABLE_NAME",
            value=infra.table_pool.table_name,
        )
        infra.table_item.grant_read_write_data(self.delete_pool.function.role)
        self.delete_pool.function.add_environment(
            key="ITEM_TABLE_NAME",
            value=infra.table_item.table_name,
        )

        self.dice = LambdaConstruct(self, "dice", infra)
        dice.add_method(
            http_method="GET",
            integration=apigw.LambdaIntegration(
                handler=self.dice.function,
            ),
        )
        assert self.dice.function.role is not None
        infra.table_pool.grant_read_data(self.dice.function.role)
        self.dice.function.add_environment(
            key="POOL_TABLE_NAME",
            value=infra.table_pool.table_name,
        )
        infra.table_item.grant_read_data(self.dice.function.role)
        self.dice.function.add_environment(
            key="ITEM_TABLE_NAME",
            value=infra.table_item.table_name,
        )
