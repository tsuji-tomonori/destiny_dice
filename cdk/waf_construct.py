from typing import Any, Self

from aws_cdk import aws_wafv2 as wafv2
from constructs import Construct

from cdk.paramater import build_name, paramater


class WafConstruct(Construct):
    def __init__(
        self: Self,
        scope: Construct,
        construct_id: str,
        **kwargs: Any,  # noqa: ANN401
    ) -> None:
        super().__init__(scope, construct_id, **kwargs)

        self.ipset = wafv2.CfnIPSet(
            scope_=self,
            id="ipset",
            addresses=paramater["waf"]["common"]["addresses"],
            ip_address_version=paramater["waf"]["common"]["ip_address_version"],
            scope=paramater["waf"][construct_id]["scope"],
            name=build_name("ipset", construct_id),
        )

        self.webacl = wafv2.CfnWebACL(
            scope_=self,
            id="webacl",
            default_action=wafv2.CfnWebACL.DefaultActionProperty(
                block=wafv2.CfnWebACL.BlockActionProperty(),
            ),
            scope=paramater["waf"][construct_id]["scope"],
            visibility_config=wafv2.CfnWebACL.VisibilityConfigProperty(
                cloud_watch_metrics_enabled=True,
                metric_name=paramater["waf"]["common"]["metric_name"],
                sampled_requests_enabled=True,
            ),
            name=build_name("webacl", construct_id),
            rules=[
                wafv2.CfnWebACL.RuleProperty(
                    name=build_name("rule", construct_id),
                    priority=paramater["waf"]["common"]["priority"],
                    statement=wafv2.CfnWebACL.StatementProperty(
                        ip_set_reference_statement=wafv2.CfnWebACL.IPSetReferenceStatementProperty(
                            arn=self.ipset.attr_arn,
                        ),
                    ),
                    visibility_config=wafv2.CfnWebACL.VisibilityConfigProperty(
                        cloud_watch_metrics_enabled=True,
                        metric_name=paramater["waf"]["common"]["metric_name"],
                        sampled_requests_enabled=True,
                    ),
                    action=wafv2.CfnWebACL.RuleActionProperty(
                        allow=wafv2.CfnWebACL.AllowActionProperty(),
                    ),
                ),
            ],
        )
