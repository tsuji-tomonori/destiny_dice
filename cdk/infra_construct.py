from typing import Any, Self

import aws_cdk as cdk
from aws_cdk import aws_dynamodb as dynamdb
from aws_cdk import aws_sns as sns
from constructs import Construct

from cdk.paramater import build_name
from cdk.waf_construct import WafConstruct


class InfraConstruct(Construct):
    def __init__(
        self: Self,
        scope: Construct,
        construct_id: str,
        **kwargs: Any,  # noqa: ANN401
    ) -> None:
        super().__init__(scope, construct_id, **kwargs)

        # DynamoDB
        self.table_pool = dynamdb.Table(
            scope=self,
            id="pool",
            partition_key=dynamdb.Attribute(
                name="pool_name",
                type=dynamdb.AttributeType.STRING,
            ),
            billing_mode=dynamdb.BillingMode.PAY_PER_REQUEST,
            removal_policy=cdk.RemovalPolicy.DESTROY,
            table_name=build_name("table", "pool"),
        )

        self.table_item = dynamdb.Table(
            scope=self,
            id="item",
            partition_key=dynamdb.Attribute(
                name="pool_name",
                type=dynamdb.AttributeType.STRING,
            ),
            sort_key=dynamdb.Attribute(
                name="item_id",
                type=dynamdb.AttributeType.NUMBER,
            ),
            billing_mode=dynamdb.BillingMode.PAY_PER_REQUEST,
            removal_policy=cdk.RemovalPolicy.DESTROY,
            table_name=build_name("table", "item"),
        )

        # AWS SNS
        self.sns_topic = sns.Topic(
            self,
            id="topic",
            topic_name=build_name("topic", "error_notify"),
        )

        self.waf = WafConstruct(self, "apigw")
