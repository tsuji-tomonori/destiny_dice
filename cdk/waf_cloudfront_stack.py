from typing import Any, Self

from aws_cdk import Stack
from constructs import Construct

from cdk.waf_construct import WafConstruct


class WafCloudFrontStack(Stack):
    def __init__(
        self: Self,
        scope: Construct | None = None,
        construct_id: str | None = None,
        **kwargs: Any,  # noqa: ANN401
    ) -> None:
        super().__init__(scope, construct_id, **kwargs)

        self.waf = WafConstruct(self, "cloudfront")
