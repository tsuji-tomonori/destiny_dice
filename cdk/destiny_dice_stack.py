from typing import Any, Self

from aws_cdk import Stack
from constructs import Construct

from cdk.app_construct import AppConstruct
from cdk.infra_construct import InfraConstruct
from cdk.static_construct import StaticConstruct


class DestinyDiceStack(Stack):
    def __init__(
        self: Self,
        scope: Construct,
        construct_id: str,
        web_acl_arn: str,
        **kwargs: Any,  # noqa: ANN401
    ) -> None:
        super().__init__(scope, construct_id, **kwargs)

        self.infra = InfraConstruct(self, "infra")
        self.static = StaticConstruct(self, "static", web_acl_arn)
        self.app = AppConstruct(self, "app", self.infra)
