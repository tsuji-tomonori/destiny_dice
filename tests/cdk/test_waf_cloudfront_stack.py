import aws_cdk as cdk
from aws_cdk import assertions
from syrupy.matchers import path_type

from cdk.waf_cloudfront_stack import WafCloudFrontStack


def test_snapshot(snapshot) -> None:
    app = cdk.Stack()
    stack = WafCloudFrontStack(app, "test")
    template = assertions.Template.from_stack(stack)

    matcher = path_type(
        {
            r".*\.S3Key$": (str,),
            r".*\.Addresses$": (list,),
        },
        regex=True,
    )
    assert template.to_json() == snapshot(matcher=matcher)
