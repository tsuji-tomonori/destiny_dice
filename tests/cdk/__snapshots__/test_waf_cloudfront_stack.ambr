# serializer version: 1
# name: test_snapshot
  dict({
    'Parameters': dict({
      'BootstrapVersion': dict({
        'Default': '/cdk-bootstrap/hnb659fds/version',
        'Description': 'Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]',
        'Type': 'AWS::SSM::Parameter::Value<String>',
      }),
    }),
    'Resources': dict({
      'cloudfrontipset4C580034': dict({
        'Properties': dict({
          'Addresses': list,
          'IPAddressVersion': 'IPV4',
          'Name': 'dice-ipset-cloudfront',
          'Scope': 'CLOUDFRONT',
        }),
        'Type': 'AWS::WAFv2::IPSet',
      }),
      'cloudfrontwebacl8A66464E': dict({
        'Properties': dict({
          'DefaultAction': dict({
            'Block': dict({
            }),
          }),
          'Name': 'dice-webacl-cloudfront',
          'Rules': list([
            dict({
              'Action': dict({
                'Allow': dict({
                }),
              }),
              'Name': 'dice-rule-cloudfront',
              'Priority': 100,
              'Statement': dict({
                'IPSetReferenceStatement': dict({
                  'Arn': dict({
                    'Fn::GetAtt': list([
                      'cloudfrontipset4C580034',
                      'Arn',
                    ]),
                  }),
                }),
              }),
              'VisibilityConfig': dict({
                'CloudWatchMetricsEnabled': True,
                'MetricName': 'destiny_dice',
                'SampledRequestsEnabled': True,
              }),
            }),
          ]),
          'Scope': 'CLOUDFRONT',
          'VisibilityConfig': dict({
            'CloudWatchMetricsEnabled': True,
            'MetricName': 'destiny_dice',
            'SampledRequestsEnabled': True,
          }),
        }),
        'Type': 'AWS::WAFv2::WebACL',
      }),
    }),
    'Rules': dict({
      'CheckBootstrapVersion': dict({
        'Assertions': list([
          dict({
            'Assert': dict({
              'Fn::Not': list([
                dict({
                  'Fn::Contains': list([
                    list([
                      '1',
                      '2',
                      '3',
                      '4',
                      '5',
                    ]),
                    dict({
                      'Ref': 'BootstrapVersion',
                    }),
                  ]),
                }),
              ]),
            }),
            'AssertDescription': "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
          }),
        ]),
      }),
    }),
  })
# ---
